sycl/doc/design/GroupSort.md

```diff
@@ -0,0 +1,220 @@
+implementations from the fallback library will be called.
+
+Interface for the library and backends:
+
+```cpp
+// for default sorting algorithm
+void __devicelib_default_work_group_joint_sort_ascending_<encoded_param_types>(T* first, uint n, byte* scratch);
+
+void __devicelib_default_work_group_joint_sort_descending_<encoded_param_types>(T* first, uint n, byte* scratch);
+
+// for fixed-size arrays
+void __devicelib_default_work_group_private_sort_close_ascending_<encoded_param_types>(T* first, uint n, byte* scratch);
+
+void __devicelib_default_work_group_private_sort_close_descending_<encoded_param_types>(T* first, uint n, byte* scratch);
+
+void __devicelib_default_work_group_private_sort_spread_ascending_<encoded_param_types>(T* first, uint n, byte* scratch);
+
+void __devicelib_default_work_group_private_sort_spread_descending_<encoded_param_types>(T* first, uint n, byte* scratch);
+
+// for sub-groups
+T __devicelib_default_sub_group_private_sort_ascending_<encoded_scalar_param_type>(T value);
+
+T __devicelib_default_sub_group_private_sort_descending_<encoded_scalar_param_type>(T value);
+
+// for key value sorting using the default algorithm
+void __devicelib_default_work_group_joint_sort_ascending_<encoded_param_types>(T* keys_first, U* values_first, uint n, byte* scratch);
+
+void __devicelib_default_work_group_joint_sort_descending_<encoded_param_types>(T* keys_first, U* values_first, uint n, byte* scratch);
+
+// for key value sorting using fixed-size arrays
+void __devicelib_default_work_group_private_sort_close_ascending_<encoded_param_types>(T* keys_first, U* values_first, uint n, byte* scratch);
+
+void __devicelib_default_work_group_private_sort_close_descending_<encoded_param_types>(T* keys_first, U* values_first, uint n, byte* scratch);
+
+void __devicelib_default_work_group_private_sort_spread_ascending_<encoded_param_types>(T* keys_first, U* values_first, uint n, byte* scratch);
+
+void __devicelib_default_work_group_private_sort_spread_descending_<encoded_param_types>(T* keys_first, U* values_first, uint n, byte* scratch);
+
+```
+
+Notes:
+- `T`, `U` are from the following list `i8`, `i16`,
+  `i32`, `i64`, `u8`, `u16`, `u32`, `u64`, `f16`, `f32`, `f64`.
+- `encoded_param_types` is `T` prepended with `p1` for global/private address
+  space and `p3` for shared local memory.
+- `first` is a pointer to the actual data for sorting.
+- The type of `n` (number of elements) is u32.
+- `keys_first` points to "keys" for key-value sorting.
+  "Keys" are comparing and moving during the sorting.
+- `scratch` is a temporary storage (local or global) that can be used by backends.
+  The type of `scratch` is always `byte*`.
+- `values_first` points to "values" for key-value sorting. "Keys" are only moving
+  corresponding the "keys" order during the sorting.
+
+Examples:
+```cpp
+void __devicelib_default_work_group_joint_sort_ascending_p1i32_u32_p3i8(int* first, uint n, byte* scratch);
```

Comment:
We have `p1`/`p3` in the "mangling" but actual operands are generic address space, is that expected?

Also, just an idea to discuss, we use the same device compiler on Lin/Win so the C++ mangling is stable. Can we use a normal C++ template with "extern template" to avoid manual mangling?